exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/map/UncheckedFlatMapProof.java

```diff
@@ -249,42 +257,114 @@ private CheckedMapProof checkProof() {
-    DbKey dbKey = DbKey.newLeafKey(entry.getKey());
-    HashCode valueHash = HASH_FUNCTION.hashByteString(entry.getValue());
-    return getSingleEntryRootHash(dbKey, valueHash);
+  private static HashCode getSingleEntryProofIndexHash(MapEntry<ByteString, ByteString> mapEntry) {
+    DbKey dbKey = DbKey.newLeafKey(mapEntry.getKey());
+    HashCode valueHash = getLeafEntryHash(mapEntry.getValue());
+    HashCode merkleRoot = getSingleEntryMerkleRoot(dbKey, valueHash);
+    return getIndexHash(merkleRoot);
   }
 
-  private static HashCode getSingleEntryRootHash(DbKey key, HashCode valueHash) {
+  private static HashCode getIndexHash(HashCode merkleRoot) {
+    return HASH_FUNCTION.newHasher()
+        .putByte(MAP_ROOT_PREFIX)
+        .putObject(merkleRoot, hashCodeFunnel())
+        .hash();
+  }
+
+  private static HashCode getSingleEntryMerkleRoot(DbKey key, HashCode valueHash) {
     assert key.getNodeType() == Type.LEAF;
     return HASH_FUNCTION.newHasher()
+        .putByte(MAP_NODE_PREFIX)
         .putObject(key, dbKeyFunnel())
         .putObject(valueHash, hashCodeFunnel())
         .hash();
   }
 
-  private static HashCode getMapEntryHash(MapEntry<ByteString, ByteString> entry) {
-    return HASH_FUNCTION.hashByteString(entry.getValue());
+  private static HashCode getLeafEntryHash(ByteString entryValue) {
+    return HASH_FUNCTION.newHasher()
+        .putByte(BLOB_PREFIX)
+        .putBytes(entryValue.toByteArray())
+        .hash();
   }
 
   private static HashCode computeBranchHash(MapProofEntry leftChild, MapProofEntry rightChild) {
+    byte[] leftChildKey = compress(leftChild.getDbKey());
+    byte[] rightChildKey = compress(rightChild.getDbKey());
     return HASH_FUNCTION
         .newHasher()
+        .putByte(MAP_NODE_PREFIX)
         .putObject(leftChild.getHash(), hashCodeFunnel())
         .putObject(rightChild.getHash(), hashCodeFunnel())
-        .putObject(leftChild.getDbKey(), dbKeyFunnel())
-        .putObject(rightChild.getDbKey(), dbKeyFunnel())
+        .putBytes(leftChildKey)
+        .putBytes(rightChildKey)
         .hash();
   }
 
+  /**
+   * Returns a LEB128 compressed binary representation of the given db key.
+   */
+  public static byte[] compress(DbKey dbKey) {
+    byte[] buffer = new byte[Hashing.DEFAULT_HASH_SIZE_BYTES * 2];
+    int bytesWritten = writeCompressed(buffer, dbKey);
+    byte[] result = new byte[bytesWritten];
+    System.arraycopy(buffer, 0, result, 0, bytesWritten);
+
+    return result;
+  }
+
+  /**
+   * Returns a number of bytes written into buffer which contains LEB128 compressed binary
+   * representation of the given db key.
+   */
+  private static int writeCompressed(byte[] buffer, DbKey dbKey) {
+    int bitsLength = dbKey.getNumSignificantBits();
+    // Perform division, rounding the result up
+    int wholeBytesLength = (bitsLength + Byte.SIZE - 1) / Byte.SIZE;
+    byte[] key = dbKey.getKeySlice();
+
+    int bytesWritten = writeUnsignedLeb128(buffer, bitsLength);
```

Comment:
I find this code/documentation rather confusing. We write up to 9 bits of information (because bitsLength is from 0 to 256) using writeUnsignedLeb128 with a fancy loop.
 Does this method only compress the number of significant bits, as it then just copies the key as is?

Reply:
I believe so, yes. I think @alekseysidorov could provide more details as the author of the original algorithm